### ISSUE TITLE
Feature request :: Right-align cursors/selections

### ISSUE THREAD

**Issue by whatsupbros** [yo1dog/vscode-cursor-align#12]
A feature to right-align all cursors/selections by adding necessary white space from the left would be really nice.

This is useful when you have several "parts" in your statement, and need to align those parts between each other, in example, `SELECT`/`FROM`/`WHERE`/`GROUP BY`/`ORDER BY` parts in SQL queries.

### Original text

```
She loves pie.
My friend loves chocolate cake.
Everybody loves muffins.
```

### Left-aligned at the second word (current functionality)

```
She       loves pie.
My friend loves chocolate cake.
Nobody    loves muffins.
```

### Right-aligned at the second word (desired functionality)

```
      She loves pie.
My friend loves chocolate cake.
   Nobody loves muffins.
```

### Real world usage example in SQL query

```
SELECT t1.field1,
       t2.field2,
       t2.field3
  FROM table1 t1,
       table2 t2
 WHERE t1.field1 = t2.field1
   AND t2.field2 = 'ABC'
   AND EXISTS ( SELECT NULL
                  FROM table3 t3
                 WHERE t3.field1 = t1.field1
                   AND t3.field2 = 123 )
```

As shown here, `SELECT`, `FROM`, `WHERE`, `AND` are right-aligned, which is required in some projects.

**Comment by yo1dog** [comment]
Can you give me some more details on how exactly you would expect this to work? I am having trouble understanding right-alignment in left-aligned environments.

How would the extension know where to insert the whitespace?

For example, what should happen in the following to "right-align" to words "love"? A or B? And how would the extension know?

Perhaps this would only be possible with selections which would define the bounds of the text to be right aligned. Outcome B would be possible if the user selected "She", "My Friend", and "Nobody".

```
var i = 0; // She loves pie.
fu();      // My friend loves chocolate cake.
bar();     // Nobody loves muffins.

A. Start of line.
      var i = 0; // She loves pie.
fu();      // My friend loves chocolate cake.
   bar();     // Nobody loves muffins.

B. ???
var i = 0; //       She loves pie.
fu();      // My friend loves chocolate cake.
bar();     //    Nobody loves muffins.

```

**Comment by whatsupbros** [comment]
Sorry for the late reply, I somehow didn't get any notifications about your comment.

> Can you give me some more details on how exactly you would expect this to work? I am having trouble understanding right-alignment in left-aligned environments.
>
> How would the extension know where to insert the whitespace?

What I had in mind was a simple implementation of it - the user places a number of cursors on several lines. Then, all lines are aligned at the cursors from both sides - from the right and from the left ones.

The current behavior left-aligns cursors, which means everything on the right from the cursors is aligned at them.

For right-aligned cursors also everything that is on the left from the cursors should be aligned at them (preserving one white-space character on the left if any were there).

As in my example:

#### Original text

Notice the "|" symbols, which represent the placed cursors

```
She |loves pie.
My friend |loves chocolate cake.
Nobody |loves muffins.
```

#### Left-aligned text at the placed cursors

Is done by adding spaces between the cursor and the text on the left of it.

```
She       |loves pie.
My friend |loves chocolate cake.
Nobody    |loves muffins.
```

#### Right-aligned text at the placed cursors

Is done by adding spaces at the beginning of each line (and removing extra white space characters between the cursor and the next non-whitespace character when necessary).

```
      She |loves pie.
My friend |loves chocolate cake.
   Nobody |loves muffins.
```
 
> For example, what should happen in the following to "right-align" to words "love"? A or B? And how would the extension know?

A.
This looks strange, because this is not a valid use-case for the right-aligned cursors.
A valid use-case should usually include a long but atomic block of code (containing multiple sections) that should be formatted, e.g. SQL queries with their `SELECT`/`FROM`/`WHERE`/`GROUP BY`/`ORDER BY` sections.

**Comment by yo1dog** [comment]
This extension is meant to provide agnostic text (actually cursor/selection) alignment functionality. Therefore we don't want to narrow acceptable use cases to only "atomic blocks of code". The user may not be writing code at all. Instead the extension should be a simple and predictable text editing tool.

I suspect these are perfectly valid use cases for right alignment as well. I've certainly right-aligned long numbers for readability before:
```
var fu1 = 700;
var fu2 = 3100;
var fu3 = 12000;
```
```
var fu1 =   700;
var fu2 =  3100;
var fu3 = 12000;
```
```
var fuThingamajig = 1;
var barThingamajig = 2;
var fubarThingamajig = 3;
```
```
var    fuThingamajig = 1;
var   barThingamajig = 2;
var fubarThingamajig = 3;
```

Always inserting whitespace at the start of the line solves only a narrow band of use cases and breaks expectations.

If a user selected all the numbers like so:
```
var fu1 = |700|;
var fu2 = |3100|;
var fu3 = |12000|;
```

Right aligning would produce the following: Whitespace added to the selections such that the start and ends are aligned.
```
var fu1 = |700  |;
var fu2 = |3100 |;
var fu3 = |12000|;
```

While your proposal for left aligning would produce the following: Whitespace added to ~the selections~ the start of each line such that ~the start and ends are aligned~ only the ends are aligned.
```
  var fu1 = |700|;
 var fu2 = |3100|;
var fu3 = |12000|;
```

That is unexpected. After all, this extension is called "cursor align" and not all the cursors/selections are aligned!

-----

Left alignment is defined as: Inserts whitespace to the left of selections until all the selection starts are aligned with the right-most selection start. Adds whitespace to the end of selections until all selections are the same length as the longest selection.

Right alignment could be defined as: Inserts whitespace to the left of selections until all the selection starts are aligned with the right-most selection start. Adds whitespace to the **front** of selections until all selections are the same length as the longest selection.

The only difference is in bold. This would mean while there is a difference between left and right aligning selections of 1 or more characters, there is no difference aligning 0-length selections.

I think this makes sense as this extension is meant to align cursors and is agnostic to the text around the cursors. That is, this extension's job is to align columns of selections. The only difference between left and right alignment is what happens within those columns.

This handles your use case (though it does require a bit more effort) as well as mine. 

I think this is the only way right alignment makes sense in this extension. Otherwise, right alignment should be handled by a different extension with different expectations (perhaps one that does not handle selections or multiple cursors per line at all).

I can see an argument for it being confusing that left and right alignment do the same thing for 0-length selections. But that argument would push me to the conclusion that we should not support right alignment at all, rather than adding special conditions. Special conditions breed unpredictability. For example, I would not want to add a rule that says something like "right aligning adds whitespace to the start of each selection unless all selections have a length of 0 in which case whitespace is added to the start of each line instead."

-----

Using your SQL example, lets say the original text and cursor placement is:
```
SELECT |t1.field1,
|t2.field2,
|t2.field3
FROM |table1 t1,
|table2 t2
WHERE |t1.field1 = t2.field1
AND |t2.field2 = 'ABC'
AND |EXISTS ( SELECT NULL
FROM table3 t3
WHERE t3.field1 = t1.field1
AND t3.field2 = 123 )
```

If you right align, it would have the same effect as left align:
```
SELECT |t1.field1,
       |t2.field2,
       |t2.field3
FROM   |table1 t1,
       |table2 t2
WHERE  |t1.field1 = t2.field1
AND    |t2.field2 = 'ABC'
AND    |EXISTS ( SELECT NULL
FROM table3 t3
WHERE t3.field1 = t1.field1
AND t3.field2 = 123 )
```

However, if you hit shift+home (in windows), it selects from each cursor to the start of the line:
```
|SELECT |t1.field1,
|t2.field2,
|t2.field3
|FROM |table1 t1,
|table2 t2
|WHERE |t1.field1 = t2.field1
|AND |t2.field2 = 'ABC'
|AND |EXISTS ( SELECT NULL
FROM table3 t3
WHERE t3.field1 = t1.field1
AND t3.field2 = 123 )
```

Now right aligning would have your desired effect as the columns of text to be aligned have been explicitly defined. All selections start at column 1 so no whitespace is inserted to the left of any selection. The longest selection is `SELECT ` with 7 characters, so whitespace is added to the **front** of each selection until they all have a length of 7:
```
|SELECT |t1.field1,
|       |t2.field2,
|       |t2.field3
|  FROM |table1 t1,
|       |table2 t2
| WHERE |t1.field1 = t2.field1
|   AND |t2.field2 = 'ABC'
|   AND |EXISTS ( SELECT NULL
FROM table3 t3
WHERE t3.field1 = t1.field1
AND t3.field2 = 123 )
````

**Comment by whatsupbros** [comment]
What you described is an extension of my original idea. 

And I completely agree that working with non-0-length selections brings many benefits, including the ability to right align text _inside_ them, which is extremely useful for right-alignment of numbers or variables names in the definition section.

I do not quite understand though why right- and left-alignment should do the same job for 0-length selections.

> _Right alignment could be defined as_: Inserts whitespace to the left of selections until all the selection starts are aligned with the right-most selection start. Adds whitespace to the **front** of selections until all selections are the same length as the longest selection.

This is good enough and if it is easier to implement it like that, it will already do the job (through using non-0-length selections).

But why instead of inserting whitespace to the left of selections we don't simply insert the same number of whitespaces to the beginning of each line for right-alignment? Everything else will stay the same (means, this part stays: _Adds whitespace to the **front** of selections until all selections are the same length as the longest selection._).

This will work, because naturally, users will either use "cursors" or "selections", which start at the same positions in lines for right-alignment.

#### Examples

```
var fu1 = |700|;
var fu2 = |3100|;
var fu3 = |12000|;
```

```
var |fuThingamajig = 1;|
var |barThingamajig = 2;|
var |fubarThingamajig = 3;|
```

```
SELECT |t1.field1,
|t2.field2,
|t2.field3
FROM |table1 t1,
|table2 t2
WHERE |t1.field1 = t2.field1
AND |t2.field2 = 'ABC'
AND |EXISTS ( SELECT NULL
FROM table3 t3
WHERE t3.field1 = t1.field1
AND t3.field2 = 123 )
```

In all these examples, the user is interested in:
- either right-align text in selections, beginnings of which are already left-aligned;
- or text before cursors, implicitly assuming that they define only the endings for the alignment, and the beginning of each line is used as the beginnings by default.

---

In other words, the difference between 0-length selections (cursors) and non-0-length selections is that for non-0-length selections also the inner text is right-aligned. No special conditions required.

Once again, thank you very much for the extension, I use it every day, and this "right-alignment" feature is the only thing that I am really missing in it.

**Comment by yo1dog** [comment]
I'm very glad you find it useful and I appreciate you taking the time to discuss improvements.

> But why instead of inserting whitespace to the left of selections we don't simply insert the same number of whitespaces to the beginning of each line for right-alignment?

> This will work, because naturally, users will either use "cursors" or "selections", which start at the same positions in lines for right-alignment.

> In all these examples, the user is interested in:
> 
> -  either right-align text in selections, beginnings of which are already left-aligned;

Not necessarily. The text may not already be left-aligned. A very plausible use case for right alignment:
```
var fu = |700|;
var fubar = |3100|;
var baz = |12000|;
```
I would want the whitespace inserted to the left of the selections:
```
var fu =    |  700|;
var fubar = | 3100|;
var baz =   |12000|;
```
Not at the start of each line:
```
   var fu = |  700|;
var fubar = | 3100|;
  var baz = |12000|;
  ```
If we always insert whitespace at the line start, the first result is impossible to achieve. If we always insert to the left of selections, both results are possible. To achieve the second result the user would right align the variable assignments first and then right align the numbers.

-----

> or text before cursors, implicitly assuming that they define only the endings for the alignment, and the beginning of each line is used as the beginnings by default.

> No special conditions required.

I would say this is a special condition. You are proposing different behavior under certain (special) conditions: when right aligning and when the selection length is 0.

alignment | selection | text to align
---|---|---
left | 1+ | selected text
left | 0 | selected text
right | 1+ | selected text
right | 0 | start of line to cursor

Further, we should avoid assumptions as much as possible. It is impossible to differentiate a cursor/0-length selection which the user intended to "align all text on this line before this point" and a 0-length selection the user intended to "align all text within this empty selection". For example, imagine a user wanting to right align the age column in this CSV:
```
name,|age|,food
Mark,|7|,pear
Mike,|1234|,banana
Death,|,apple
Bob,|88|,cake
```

Because Death has no age, there is a 0-length selection on line 4. If we were to use your assumption and rule about 0-length selections, we would end up with:
```
name, |   age|,food
Mark, |     7|,pear
Mike, |  1234|,banana
       Death,|,apple
Bob,  |    88|,cake
```
rather than the desired:
```
name, | age|,food
Mark, |   7|,pear
Mike, |1234|,banana
Death,|    |,apple
Bob,  |  88|,cake
```

-----

You could change your rule to apply only when **all** selections are 0-length, but we would still be making assumptions and we would still have the special condition:

alignment | selections | text to align
---|---|---
left | some 1+ | selected text
left | all 0 | selected text
right | some 1+ | selected text
right | all 0 | start of line to cursor

If we were to adopt that rule, imagine the user's confusion if they accidentally and unknowingly selected the `E` in `EXISTS`. The rules would swap from "right, all 0" to "right, some 1+", the result would very different, and the user wouldn't know why:
```
SELECT |t1.field1,
|t2.field2,
|t2.field3
FROM |table1 t1,
|table2 t2
WHERE |t1.field1 = t2.field1
AND |t2.field2 = 'ABC'
AND |E|XISTS ( SELECT NULL
FROM table3 t3
WHERE t3.field1 = t1.field1
AND t3.field2 = 123 )
```
```
SELECT  t1.field1,
        t2.field2,
        t2.field3
FROM    table1 t1,
        table2 t2
WHERE   t1.field1 = t2.field1
AND     t2.field2 = 'ABC'
AND    EXISTS ( SELECT NULL
FROM table3 t3
WHERE t3.field1 = t1.field1
AND t3.field2 = 123 )
```

"Why is right align not working!?"
"It is working but you accidentally selected a single character which completely changes how the extension handles alignment."

**Comment by whatsupbros** [comment]
Once again, the implementation that you're proposing is more than valid and it would be great to have it in the extension just like that.

---

But just to be completely on the same page, let me try to eliminate the confusion.

And the confusion lays here - this is not what I meant in my previous message:

alignment | selections | text to align
-- | -- | --
left | some 1+ | selected text
left | all 0 | selected text
right | some 1+ | selected text
right | all 0 | start of line to cursor

What I meant was:

alignment | selections | text to align
-- | -- | --
left | some 1+ | selected text
left | all 0 | selected text
right | some 1+ | start of line to cursor and selected text
right | all 0 | start of line to cursor and selected text (but since selected text is empty, this part may be omitted here)

That is why I said that no special condition required.

The only reason why I suggested it like this is because in all other applications, where users have the ability to "right-align" text, this means that the text sticks to the right border of the manipulated block - in Word or similar text editors this is the right border of the page, in other apps, where there's a "block of text" component (i.e. Paint, Draw.io, etc.), this is the right border of this block.

And instinctively, when I place a number of cursors and perform "right-alignment", I expect that first - these cursors form a line (become aligned), and then the text on the left of them will stick to this line.

And your example:

```
name,|age|,food
Mark,|7|,pear
Mike,|1234|,banana
Death,|,apple
Bob,|88|,cake
```

Would become:

```
 name,| age|,food
 Mark,|   7|,pear
 Mike,|1234|,banana
Death,|    |,apple
  Bob,|  88|,cake
```

In fact, "left-alignment" does the same - first the cursors form a line (align), then "selected text" left-aligns, and finally "everything to the right of the cursors/selections" also left-aligns.
This is just done implicitly and is so obvious, that we don't even think about it, as we live in the left-aligned word by default. However, for some right-aligned languages (i.e. Arabic or Hebrew), the situation may be not as obvious.

---

Nevertheless, this is only my take of it, and yours is as valid as mine.

**Comment by yo1dog** [comment]
Again I appreciate the discourse. I just want to make sure I understand you and that I am not missing use cases or something else.

> in other apps, where there's a "block of text" component (i.e. Paint, Draw.io, etc.), this is the right border of this block.

Agreed. The issue in our case is figuring out how the exactly the "block of text" is defined. Is it all text to the left of cursors? Or is it only selected text? I believe you are arguing for the former and I am arguing for the later.

Sounds like your proposal for the right alignment definition is (compared to left alignment): Inserts whitespace [~to the left of selections~ at the start of each line] until all the selection starts are aligned with the right-most selection start. Adds whitespace to the [~end~ front] of selections until all selections are the same length as the longest selection.

Hopefully I got that right.

That makes a lot of sense. And it certainly provides a better result in uses cases such as your original SQL example.

By my main issue still stands in that this implementation makes it impossible to solve this use case:
```
var fu = |700|;
var fubar = |3100|;
var baz = |12000|;
```
```
var fu =    |  700|;
var fubar = | 3100|;
var baz =   |12000|;
```

This is because that definition prohibits the ability to fully define the bounds of the "block of text" that should be aligned (in code I refer to these as "columns"). Instead, it always defines "start of line to cursor" as the first block to right align, and then the user's selections as a second block to right align. With left alignment, only the selections define the singular block to be aligned (ignoring multiple selections per line for the moment).

My other issue is the the fact that, while perhaps you wouldn't call it a "special condition" because it is consistent for all right alignments, the behavior of right alignment is still different enough from left alignment that it would break expectations. For example, if a user hit left align and got:
```
name, |age |,food
Mark, |7   |,pear
Mike, |1234|,banana
Death,|    |,apple
Bob,  |88  |,cake
```
Hit undo, and then hit right align, I think they would expect:
```
name, | age|,food
Mark, |   7|,pear
Mike, |1234|,banana
Death,|    |,apple
Bob,  |  88|,cake
```

That is, the same affected block of text just aligned to the right instead of the left.

**Comment by yo1dog** [comment]
Somewhat related: It is not possible to right align using tabs. So right alignment would be available with spaces only. Otherwise, you would get into an ugly mix of tabs and spaces. Another potentially broken expectation:

```
→   |select•distinct|•col
→   |from|•mytable
→   |where|•a•>•b
→   |and|•a•<•c
```
Tabs used when possible:
```
→   |select•distinct→|•col
→   |→   →   •••from→|•mytable
→   |→   →   ••where→|•a•>•b
→   |→   →   →   and→|•a•<•c
```
Or only spaces are used for left side:
```
→   |select•distinct→|•col
→   |•••••••••••from→|•mytable
→   |••••••••••where→|•a•>•b
→   |••••••••••••and→|•a•<•c
```

I think both suck.

**Comment by yo1dog** [comment]
I started implementing this and now I understand why it has been so difficult for me to put a bow on this discussion.

We have been talking about left vs right aligning text.

This extension does not align text. This extension aligns cursors. It does not care about the text at all. It does not align to the "left" nor to the "right". It simply aligns cursors into columns.

The practical effect and purpose of cursor alignment is that by aligning the cursors, the surrounding text also becomes aligned in some way that is significant to the user. But the extension doesn't know nor care about that and should not try to guess at what that significance may be.

For example, there should be no expectation that this extension would align this selected **text**. The **cursors** are already aligned so the extension will do nothing.
```
a |1234| b
c | 56 | d
e |  7 | f
```

So when we have been talking "right alignment", we have really been discussing adding a feature that allows whitespace padding to be inserted before selections rather than after. The practical effect being that adding the padding before selections could allow users to right align text in certain situations.

In this context, I can easily see this extension providing a "Align Cursors (Pad Front)" command to complement "Align Cursors". This removes all confusion and assumptions about what exactly "right alignment" could mean because we don't call it that, it doesn't pretend to do that, and the name states exactly what it does: inserts the whitespace padding at the front of selections instead of the end.

Unfortunately, I think I am still against a command like "Align Cursors (Pad Line Start)" which would insert whitespace at the start of the line. The reason now is simply because it is not possible to support multiple selections per line. If you are only adding the whitespace padding to the start of the line, it is not possible to align multiple columns because you need padding between columns to account for the variable space between columns. Example:
```
name,|age|,food,|drink|
Mark,|7|,pear,|coke|
Mike,|1234|,banana,|pepsi|
Death,|,apple,|water|
Bob,|88|,cake,|apple juice|
```
If you align the age column first, then drink:
```
 name,| age|,food,|drink|
 Mark,|   7|,pear,|coke|
 Mike,|1234|,banana,|pepsi|
Death,|    |,apple,|water|
  Bob,|  88|,cake,|apple juice|

   name,| age|,food,|      drink|
   Mark,|   7|,pear,|       coke|
 Mike,|1234|,banana,|      pepsi|
 Death,|    |,apple,|      water|
    Bob,|  88|,cake,|apple juice|
```
Vice-versa:
```
   name,|age|,food,|      drink|
     Mark,|7|,pear,|       coke|
Mike,|1234|,banana,|      pepsi|
     Death,|,apple,|      water|
     Bob,|88|,cake,|apple juice|

     name,| age|,food,|      drink|
     Mark,|   7|,pear,|       coke|
     Mike,|1234|,banana,|      pepsi|
    Death,|    |,apple,|      water|
      Bob,|  88|,cake,|apple juice|
```

We could have a completely different command like "Align Cursor Line Ends" which could be defined as: Inserts spaces at the start of lines until the right most selection end of each line is aligned with the right-most selection end. I believe is the functionality you are asking for. But it feels very different from the other commands and wouldn't result in all the cursors being aligned. Perhaps it could be its own extension. Or maybe it's fine to just throw in.

**Comment by yo1dog** [comment]
I decided to just throw it in.

As of v2.1.0 you can use the `Align Cursors (Pad Front)` command to add spaces to the front of selections (my use cases) and the `Align Line Last Cursors` command to add spaces to the front of lines to align the last cursors on each line (your SQL use case).

Enjoy! If this does not address your needs please let me know. Otherwise, closing for now.
